afqinsight/cnn.py

```diff
@@ -0,0 +1,548 @@
+        else:
+            self.n_nodes = n_nodes
+
+        # checking n_channels is passed as int
+        if not isinstance(n_channels, int):
+            raise TypeError("Parameter n_channels must be an integer.")
+        else:
+            self.n_channels = n_channels
+
+        # checking layers is passed as int
+        if not isinstance(layers, int):
+            raise TypeError("Parameter layers must be an integer.")
+        else:
+            self.layers = layers
+
+        # checking max epochs is passed as int
+        if not isinstance(max_epochs, int):
+            raise TypeError("Parameter max_epochs must be an integer.")
+        else:
+            self.max_epochs = max_epochs
+
+        if not isinstance(batch_size, int):
+            raise TypeError("Parameter batch_size must be an integer.")
+        else:
+            self.batch_size = batch_size
+
+        # checking tiner is passed as str or None
+        if not isinstance(tuner, str) and tuner is not None:
+            raise TypeError("Parameter tuner must be str.")
+        else:
+            # tuner can be None (no tuning) BayesianOptimization, Hyperband, or RandomSearch
+            self.tuner = tuner
+
+        # checking val split is passed as float
+        if not isinstance(test_size, float):
+            raise TypeError("Parameter test_size must be a float.")
+        else:
+            self.test_size = test_size
+
+        # checking strategy is passed as str and has value of 'median', 'mean', or 'knn'
+        if not isinstance(strategy, str):
+            raise TypeError("Parameter strategy must be a string.")
+        elif strategy not in ["median", "mean", "knn"]:
+            raise ValueError("Parameter strategy must be 'median', 'mean', or 'knn'.")
+        else:
+            self.strategy = strategy
+
+        if not isinstance(random_state, int):
+            raise TypeError("Parameter random_state must be an int.")
+        else:
+            self.random_state = random_state
+
+        self.directory = directory
+        self.project_name = project_name
+        self.tuner_kwargs = tuner_kwargs
+        self.model_ = None
+        self.best_hps_ = None
+
+    def _preprocess(self, X, y=None):
+        """Convert feature matrix for input into a CNN.
+
+        Masks NAN values for X and y (if y is given), imputes X, and reshapes X
+        to be in proper form for CNN model. In more conventionall machine
+        learning, X has shape (n_samples, n_features), where n_features is
+        n_nodes * n_bundles * n_metrics. However, in our CNN approach, we treat
+        each bundle/metric combination as a separate channel, analogous to RGB
+        channels in a 2D image. The remaining one dimension is the nodes
+        dimension. Thus the output has shape (n_samples, n_channels, n_nodes),
+        where n_channels = n_metrics * n_bundles.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
```

Comment:
I think we want this docstring to be a bit more clear. This is the output of afqi.load_data, so I think that this should be fine:

```suggestion
        X : array-like of shape (n_samples, n_metrics * n_nodes)
```